user-guide: add interventions section (#723)
cylc/cylc-doc: src/user-guide/interventions/index.rst

Interventions
=============

Sometimes things don't go to plan!

So Cylc allows you to take manual control of your workflow whilst it's running
allowing you to do things like edit a task's configuration, re-run a section
of your graph or override task outputs.

This section of the documentation covers some of the common interventions you
might want to perform in various scenarios.

You can perform these interventions in multiple ways:

.. tab-set::

   .. tab-item:: GUI
      :sync: gui

      The Cylc web app (GUI) can be launched with the ``cylc gui`` command.

      Some sites provide a central deployment for you where you log in via the

>       web browser rather than launching it youself.

Author: ```suggestion
      web browser rather than launching it yourself.
```

   .. tab-item:: With the Tui
      :sync: tui

>       Cylc Tui is an interactive in-terminal application, kinda like a minimal

Reviewer: Excessive informality?
```suggestion
      Cylc Tui is an interactive in-terminal application, like a minimal
```

.. _interventions.set-task-outputs:

Set Task Outputs
----------------

:Example:
   My task failed, I've gone and fixed the problem, now I want to tell Cylc
   that the task succeeded.

:Solution:
   Set the "succeeded" :term:`output <task output>` on the failed task.

.. tab-set::

   .. tab-item:: GUI
      :sync: gui

      .. image:: set-task-outputs.gui.gif
         :width: 75%

      By default this sets the "succeeded" output, press the pencil icon next
      to the trigger command to specify a different output.

   .. tab-item:: Tui
      :sync: tui

      .. image:: set-task-outputs.tui.gif
         :width: 75%

      By default, this sets the "succeeded" output. Use ``cylc set --output``
      to specify a different output.

   .. tab-item:: CLI
      :sync: cli

      .. code-block:: console

         $ cylc set <workflow>//<task>

      By default, this sets the "succeeded" output. Use the ``--output`` option
      to specify a different output.

You can also :ref:`set task prerequisites <interventions.set-task-prerequisites>`.
Should I set task outputs or prerequisites?

* If you set a task's outputs, then tasks downstream of it may start to run.

> * If you set a task's prerequisites, this will not happen.

Reviewer 2: ```suggestion
* If you set a task's prerequisites, the task itself may start to run.
```
(In both cases, we could perhaps say exactly what "may" means, but I quite like this minimalist description).

Reviewer: The question this statement would raise in my mind is:

Does that mean that if you set a tasks prerequisites it may run, but will not automatically trigger downstream tasks?

That's the question I would expect a user reading this to ask, both with and without Hilary's edits.

Reviewer 2: I think it's reasonable as-is (with my edits), because downstream tasks won't run immediately in the set-prerequisite case. The target task has to run first and generate the outputs, and even that may not happen immediately (might be other prerequisites).

               # set prereqs "1/b2:succeeded" and "1/c:succeeded" on 1/z1
               # set prereqs "all" on 1/z2
           """
   
   [runtime]
       [[b1]]
           script = sleep 600
       [[c]]
           script = false

:Example:
   A task is not running yet, because one or more of its
   :term:`prerequisites <trigger>` are not satisfied (e.g. upstream tasks
   have not run yet or have failed). I want it to ignore one or more of these
   prerequisites.

:Solution:
   Set the task's prerequisites as satisfied.

.. note::

   If you want the task to run right away, then trigger it using the same
   intervention as :ref:`interventions.re-run-a-task`.

.. tab-set::

   .. tab-item:: GUI
      :sync: gui

      .. image:: set-task-prerequisites.gui.gif
         :width: 75%

   .. tab-item:: CLI
      :sync: cli

      .. code-block:: console

         $ cylc set \
         >   --pre <prereq-cycle>/<prereq-task>:<prereq-output> \
         >   <workflow>//<cycle><task>

You can also :ref:`set task outputs <interventions.set-task-outputs>`.
Should I set task outputs or prerequisites?

* If you set a task's outputs, then tasks downstream of it may start to run.

> * If you set a task's prerequisites, this will not happen.

Reviewer 2: ```suggestion
* If you set a task's prerequisites, the task itself may start to run.
```

Set a Switch Task
-----------------

.. workflow config:

   [scheduling]
       cycling mode = integer
       initial cycle point = 1
       runahead limit = P1
       [[graph]]
           P1 = """
               start => switch
               switch:normal? => normal
               switch:alternate? => alternate

>                normal | alternate => end

Reviewer 2: This has the problem that @MetRonnie and I flagged on the optional-outputs-extension PR. It looks nice, but novice users might not understand that the `normal` thing in `switch:normal` is not the same thing as in `=> normal`.  

Not critical though, can punt it to https://github.com/cylc/cylc-flow/issues/6085

Author: IMO, it's clearer this way than the alternative but open to reasonable suggestions.

Reviewer 2: No need to change it on this PR anyway.

      switch:normal? => normal
   
      # but it can follow an alternate route if desired
      switch:alternate? => alternate
   
      normal | alternate => end

:Solution:
   Set "succeeded" as well as the desired output(s) on the task and use the
   ``-wait`` option. Cylc will follow the desired path when it gets there.

   .. note::
   
      We set the "succeeded" output to prevent the task from being re-run when the
      flow approaches it.

.. tab-set::

   .. tab-item:: GUI
      :sync: gui

      .. image:: set-a-switch-task.gui.gif
         :width: 75%

   .. tab-item:: CLI
      :sync: cli

      .. code-block:: console

         $ cylc set --wait --output=succeeded,alternate <workflow>

.. _interventions.edit-a-tasks-configuration:

Edit a Task's Configuration and Re-Run It
-----------------------------------------

:Example:
   I'm developing a task and need to be able to quickly make changes and re-run
   the task.

:Solution:
   Edit the task's definition and trigger it.

.. tab-set::

   .. tab-item:: GUI
      :sync: gui

      .. image:: edit-a-tasks-configuration.gui.gif
         :width: 75%

      .. note::

>          Any changes you make, apply only to this one instance of the task, not

Reviewer 2: ```suggestion
         Any changes you make apply only to this one instance of the task, not
```

         to any future instances.

         To change future instances, either use "broadcast" or see
         :ref:`interventions.edit-the-workflow-configuration`.

   .. tab-item:: CLI
      :sync: cli

      .. code-block:: console

         $ cylc broadcast <workflow> -p <cycle> -n <task> -s 'script=true'
         $ cylc trigger <workflow>//<cycle>/<task>

.. the "|" character adds some vertical whitespace

|

.. _interventions.edit-the-workflow-configuration:

Edit The Workflow Configuration
-------------------------------

:Example:
   I want to change the configuration of multiple tasks or add/remove tasks

>    without stopping the workflow or starting it from the beginning again.

Reviewer 2: This mentions adding or removing tasks from the graph, which has some caveats, so I think we should link here to the "Restarting or reloading after graph changes" section.

Author: Added a note.
